File: leetcode/n-ary-tree-level-order-traversal/n-ary-tree-level-order-traversal.rb
# LeetCode #429 - N-ary Tree Level Order Traversal
# https://leetcode.com/problems/n-ary-tree-level-order-traversal/

# Definition for a Node.
# class Node
#     attr_accessor :val, :children
#     def initialize(val)
#         @val = val
#         @children = []
#     end
# end

# @param {Node} root
# @return {List[List[int]]}
def levelOrder(root)
  out = []
  queue = []
  queue << root if root

  until queue.empty?
    row = []

    queue.size.times do
      node = queue.shift
      row << node.val
      queue.concat(node.children)
    end

    out << row
  end

  out
end

File: leetcode/n-ary-tree-level-order-traversal/n-ary-tree-level-order-traversal.java
// LeetCode #429 - N-ary Tree Level Order Traversal
// https://leetcode.com/problems/n-ary-tree-level-order-traversal/

/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<List<Integer>> levelOrder(Node root) {
        List<List<Integer>> rows = new ArrayList<>();
        if (root == null) return rows;

        Queue<Node> queue = new ArrayDeque<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            List<Integer> row = new ArrayList<>();
            rows.add(row);

            for (var breadth = queue.size(); breadth > 0; --breadth) {
                var node = queue.remove();
                row.add(node.val);
                queue.addAll(node.children);
            }
        }

        return rows;
    }
}
